File: frontend/types/finance.ts
export type FinanceOverview = {
  revenue: number | null;
  cash_collected: number | null;
  refunds_paid: number | null;
  expenses: number | null;
  receivables: number | null;
  payables: number | null;
  cash_in: number | null;
  cash_out: number | null;
  net_operating: number | null;
};

export type FinanceTransactionDirection = 'in' | 'out';
export type FinanceTransactionStatus = 'posted' | 'pending' | 'reversed' | 'completed' | 'paid' | 'unpaid' | 'partial';
export type FinanceTransactionOriginType =
  | 'sale_fulfillment'
  | 'return_refund'
  | 'manual_payment'
  | 'manual_expense';
export type FinanceCounterpartyType = 'customer' | 'vendor' | 'internal';

export type FinanceReport = {
  from_date: string;
  to_date: string;
  expense_total: number;
  expense_trend: Array<{ period: string; amount: number }>;
  receivables_total: number;
  payables_total: number | null;
  net_operating_snapshot: number | null;
  deferred_metrics: Array<{ metric: string; reason: string }>;
};

export type Expense = {
  expense_id: string;
  expense_date: string;
  category: string;
  amount: number;
  payment_status: 'paid' | 'unpaid' | 'partial';
  note: string;
  created_at: string;
  updated_at: string;
};

export type Receivable = {
  sale_id: string;
  sale_no: string;
  customer_id?: string | null;
  customer_name: string;
  sale_date: string;
  grand_total: number;
  amount_paid: number;
  outstanding_balance: number;
  payment_status: string;
};

export type Payable = {
  transaction_id: string;
  reference: string;
  vendor_name: string;
  origin_type: FinanceTransactionOriginType;
  occurred_at: string;
  amount: number;
  status: string;
  note: string;
};

export type FinanceReceivable = Receivable;
export type FinancePayable = Payable;

export type FinanceTransaction = {
  transaction_id: string;
  occurred_at: string;
  origin_type: FinanceTransactionOriginType;
  origin_id: string | null;
  direction: FinanceTransactionDirection;
  status: FinanceTransactionStatus;
  currency_code: string;
  amount: number;
  reference: string;
  note: string;
  counterparty_type: FinanceCounterpartyType | null;
  counterparty_id: string | null;
  counterparty_name: string;
  finance_posted_at?: string | null;
  editable: boolean;
  source_label: string;
};

export type FinanceTransactionList = {
  transactions: FinanceTransaction[];
  total: number;
  limit: number;
  offset: number;
};

export type FinanceWorkspace = {
  overview: FinanceOverview;
  commerce_transactions: FinanceTransaction[];
  manual_transactions: FinanceTransaction[];
  recent_refunds: FinanceTransaction[];
  receivables: Receivable[];
  payables: Payable[];
};

export type FinanceTransactionInput = {
  origin_type: 'manual_payment' | 'manual_expense';
  occurred_at?: string;
  amount: number;
  direction: FinanceTransactionDirection;
  status?: FinanceTransactionStatus;
  currency_code?: string;
  reference: string;
  note: string;
  counterparty_name?: string;
  counterparty_type?: FinanceCounterpartyType;
};

export type FinanceRefundInput = {
  refund_date?: string;
  amount: number;
  method: string;
  reference: string;
  note: string;
};

File: easy_ecom/api/schemas/finance.py
from __future__ import annotations

from typing import List, Literal, Optional
from pydantic import BaseModel, Field

TransactionType = Literal["payment", "expense"]
PaymentDirection = Literal["in", "out"]
PaymentStatus = Literal["paid", "unpaid", "partial", "completed", "pending", "succeeded"]


class FinanceTransaction(BaseModel):
    entry_id: str
    entry_date: str
    entry_type: TransactionType
    direction: PaymentDirection
    category: str
    amount: float
    reference: str
    note: str
    payment_status: Optional[PaymentStatus] = None
    vendor_name: Optional[str] = None


class TransactionListResponse(BaseModel):
    transactions: List[FinanceTransaction]
    total: int
    limit: int
    offset: int


class CreateTransactionRequest(BaseModel):
    entry_type: TransactionType = Field(..., description="Type of transaction: 'payment' or 'expense'")
    entry_date: Optional[str] = Field(None, description="Date of transaction (ISO format)")
    category: str = Field(..., description="Category of transaction")
    amount: float = Field(..., gt=0, description="Amount of transaction")
    direction: PaymentDirection = Field(..., description="Direction: 'in' for income, 'out' for expense")
    reference: str = Field("", description="Reference number or identifier")
    note: str = Field("", description="Additional notes")
    vendor_name: Optional[str] = Field(None, description="Vendor name (for expenses)")
    payment_status: Optional[PaymentStatus] = Field(None, description="Payment status (for expenses)")


class UpdateTransactionRequest(BaseModel):
    entry_type: TransactionType = Field(..., description="Type of transaction: 'payment' or 'expense'")
    entry_date: Optional[str] = Field(None, description="Date of transaction (ISO format)")
    category: Optional[str] = Field(None, description="Category of transaction")
    amount: Optional[float] = Field(None, gt=0, description="Amount of transaction")
    direction: Optional[PaymentDirection] = Field(None, description="Direction: 'in' for income, 'out' for expense")
    reference: Optional[str] = Field(None, description="Reference number or identifier")
    note: Optional[str] = Field(None, description="Additional notes")
    vendor_name: Optional[str] = Field(None, description="Vendor name (for expenses)")
    payment_status: Optional[PaymentStatus] = Field(None, description="Payment status (for expenses)")


class FinanceOverviewResponse(BaseModel):
    sales_revenue: Optional[float] = None
    expense_total: Optional[float] = None
    receivables: Optional[float] = None
    payables: Optional[float] = None
    cash_in: Optional[float] = None
    cash_out: Optional[float] = None
    net_operating: Optional[float] = None


class FinanceReceivable(BaseModel):
    sale_id: str
    sale_no: str
    customer_id: Optional[str] = None
    customer_name: str
    sale_date: str
    grand_total: float
    amount_paid: float
    outstanding_balance: float
    payment_status: str


class FinancePayable(BaseModel):
    expense_id: str
    expense_number: str
    vendor_name: str
    category: str
    expense_date: str
    amount: float
    payment_status: str
    note: str


class FinanceWorkspaceResponse(BaseModel):
    overview: FinanceOverviewResponse
    transactions: List[FinanceTransaction]
    receivables: List[FinanceReceivable]
    payables: List[FinancePayable]
